### PR TITLE
API-6603 - Command handling now returns structured error types

### PR DESCRIPTION
**Pull request by AndySpaven**
Type changes make up the bulk of the changes

### CODE REVIEW

File: app/uk/gov/hmrc/thirdpartyapplication/controllers/ApplicationCommandController.scala [hunk 1]
```diff
@@ -31,8 +29,24 @@ import uk.gov.hmrc.thirdpartyapplication.domain.models.{ApplicationCommand, Appl
...
+import uk.gov.hmrc.thirdpartyapplication.services.commands.CommandHandler
+import uk.gov.hmrc.thirdpartyapplication.services.commands.CommandFailures._
 
 object ApplicationCommandController {
+  case class DispatchRequest(command: ApplicationCommand, verifiedCollaboratorsToNotify: Set[LaxEmailAddress])
+
+  object DispatchRequest {
+    import ApplicationCommandFormatters._
+
+    val readsExactDispatchRequest: Reads[DispatchRequest] = Json.reads[DispatchRequest]
```
Comment (johnsgp) [inline app/uk/gov/hmrc/thirdpartyapplication/controllers/ApplicationCommandController.scala:44]: What is an "exact" dispatch request or command? Does this need a comment or slightly different name? Does it mean it's only one or the other, so maybe "pure" is a better word?

File: app/uk/gov/hmrc/thirdpartyapplication/controllers/ApplicationCommandController.scala [hunk 2]
```diff
@@ -51,35 +65,46 @@ class ApplicationCommandController @Inject() (
...
-  import uk.gov.hmrc.thirdpartyapplication.services.commands.CommandHandler.CommandSuccess
-
-  private def fails(applicationId: ApplicationId)(e: NonEmptyChain[String]) = {
-    logger.warn(s"Command Process failed for $applicationId because ${e.toList.mkString("[", ",", "]")}")
-    BadRequest("Failed to process command")
+  
+  private def fails(applicationId: ApplicationId)(e: CommandHandler.Failures) = {
+
+    val details = e.toList.map( _ match {
+      case _ @ ApplicationNotFound => "Application not found"
```
Comment (johnsgp) [inline app/uk/gov/hmrc/thirdpartyapplication/controllers/ApplicationCommandController.scala:77]: I think scalafmt will reorganise the arrow

File: app/uk/gov/hmrc/thirdpartyapplication/controllers/Model.scala [hunk 3]
```diff
@@ -46,6 +46,7 @@ case class RejectUpliftRequest(gatekeeperUserId: String, reason: String)
 
 case class ResendVerificationRequest(gatekeeperUserId: String)
 
+@deprecated("Bobbins", "immediately")
```
Comment (johnsgp) [inline app/uk/gov/hmrc/thirdpartyapplication/controllers/Model.scala:49]: Could this be more descriptive?

File: app/uk/gov/hmrc/thirdpartyapplication/services/ApplicationService.scala [hunk 4]
```diff
@@ -527,13 +527,13 @@ class ApplicationService @Inject() (
       newRedirectUris,
       timestamp = LocalDateTime.now(clock)
     )
-
-    applicationCommandDispatcher.dispatch(applicationId, updateRedirectUris)
+    println(updateRedirectUris)
```
Comment (johnsgp) [inline app/uk/gov/hmrc/thirdpartyapplication/services/ApplicationService.scala:530]: Please remove println

File: shared-test/uk/gov/hmrc/thirdpartyapplication/util/ActorTestData.scala [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+    def actor() = Actors.GatekeeperUser(userName)
+  }
+
+  implicit class ActorLaxEmailSyntax(email: LaxEmailAddress) {
+    def actor() = Actors.AppCollaborator(email)
+  }
+
+  implicit class ActorCollaboratorSyntax(collaborator: Collaborator) {
+    def actor() = Actors.AppCollaborator(collaborator.emailAddress)
+  }
```
Comment (johnsgp) [inline shared-test/uk/gov/hmrc/thirdpartyapplication/util/ActorTestData.scala:35]: These names feel Hungarian notationish. How about ActorByName, ActorByEmail, ActorByCollaborator?

File: app/uk/gov/hmrc/thirdpartyapplication/services/ApplicationService.scala (hunk 4 above)
Comment (AndySpaven) [reply app/uk/gov/hmrc/thirdpartyapplication/services/ApplicationService.scala:530]: Done (and some others)

File: app/uk/gov/hmrc/thirdpartyapplication/controllers/Model.scala (hunk 3 above)
Comment (AndySpaven) [reply app/uk/gov/hmrc/thirdpartyapplication/controllers/Model.scala:49]: Okay

File: app/uk/gov/hmrc/thirdpartyapplication/controllers/ApplicationCommandController.scala (hunk 1 above)
Comment (AndySpaven) [reply app/uk/gov/hmrc/thirdpartyapplication/controllers/ApplicationCommandController.scala:44]: Pure has connotations that don't apply here IMO.  Let me see if I can reword there two readers.

File: shared-test/uk/gov/hmrc/thirdpartyapplication/util/ActorTestData.scala (hunk 5 above)
Comment (AndySpaven) [reply shared-test/uk/gov/hmrc/thirdpartyapplication/util/ActorTestData.scala:35]: These follow a kind of "standard" approach to Syntax as conversions (implicit conversions).  They are never referenced by their names.

File: app/uk/gov/hmrc/thirdpartyapplication/controllers/ApplicationCommandController.scala (hunk 1 above)
Comment (johnsgp) [reply app/uk/gov/hmrc/thirdpartyapplication/controllers/ApplicationCommandController.scala]: I like the new names.

**Review by johnsgp — APPROVED** [review]

**Pull request merged**